### PR TITLE
chore: Workspace as an object in the SDK.

### PR DESCRIPTION
**Pull request by wes-turner**
## Description

New class for workspaces within the SDK.

Create them from IDs! Create them from names! Create all the workspaces!

Mostly just a stub class at this point.

## Test Plan

Create a Workspace in the SDK with the constructor `determined.common.experimental.workspace.Workspace`. Call `reload` on it. There isn't much functionality in workspaces yet and not much to see, but if you figure out how to change the workspace's name then reloading will populate that to the client.

Also, maybe test constructing the Workspace via names that do/don't exist. Can master have two different workspaces with the same name (I don't think it can)? What would happen?

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/common/experimental/workspace.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    def reload(self) -> None:
+        assert not (self.id is None and self.name is None)
+        if self.id is None:
+            resp = bindings.get_GetWorkspaces(session=self._session, name=self.name)
+            if len(resp.workspaces) == 0:
+                raise ValueError(f"No workspace found with name {self.name}")
+            workspace_binding = resp.workspaces[0]
+        else:
+            resp = bindings.get_GetWorkspace(session=self._session, id=self.id)
+            workspace_binding = resp.workspace
```
Comment (mpkouznetsov) [inline harness/determined/common/experimental/workspace.py:50]: singular or plural? I see `workspace_bindings` elsewhere.

Comment (wes-turner) [reply harness/determined/common/experimental/workspace.py]: fixed!

File: harness/determined/common/experimental/workspace.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        self.name = workspace_bindings.name
+
+    def reload(self) -> None:
+        resp = bindings.get_GetWorkspace(session=self._session, id=self.id)
+        workspace_bindings = resp.workspace
+
+        self._hydrate(workspace_bindings)
+
+
+def _get_workspace_id_from_name(session: api.Session, name: str) -> int:
```
Comment (azhou-determined) [inline harness/determined/common/experimental/workspace.py:61]: is there a reason why this method is separate from the class? do we envision callers of this method from outside the Workspace class?

Comment (wes-turner) [reply harness/determined/common/experimental/workspace.py:61]: All other things being equal, I think class methods are harder to reason about than module-level functions. You can be reasonably confident module-level functions have no side effects and you know exactly what the relevant parameters are.

And like I know that classes exist for a reason. Maybe three of them are:
1. Persisting side-effects
2. Cutting down on typing when passing parameters (or just syntactic sugar where `fun(Obj) -> Obj.fun()`)
3. Organizing namespaces

I'm fully in support of (1) when it's needed. Side effects are hard, but if they're convenient to your problem then it's dope to be able to put them into an object. (2)... sometimes I like it, sometimes I don't? Whatever makes the code easier to read and think about, I guess, depending on how much brevity you're gaining for the sacrifice to explicitness. And at least in Python, I think the community has decided that (3) is an antipattern -- modules are this language's preferred organizational tool.

This means that when (1) isn't an issue, I generally prefer to stick helper functions outside the class and directly into the module namespace.

I dunno. This is a preference I picked up somewhere in my last job. If you think it's gross (even just a little) or I've missed other reasons to prefer class-level methods, I'll do my best to talk about it with an open mind.

Comment (azhou-determined) [reply harness/determined/common/experimental/workspace.py:61]: that makes sense. the method doesn't really require persisting of any side-effects, or "state", so it doesn't need to be a method of the class. i was thinking along the lines of 3. organizing namespaces, but you are right that it doesn't seem very pythonic. thanks for the detailed explanation.

File: harness/tests/determined/common/experimental/test_workspace.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+from determined.common import api
+from determined.common.api import bindings
+from determined.common.experimental import workspace
+from tests.fixtures import api_responses
+
+_MASTER = "http://localhost:8080"
+
+
+@pytest.fixture
+def standard_session() -> api.Session:
```
Comment (azhou-determined) [inline harness/tests/determined/common/experimental/test_workspace.py:13]: not really a comment for this PR, but i wonder if these common fixtures like `standard_session` and `_MASTER` should be defined elsewhere, like in a shared utils or something. it seems like we're using them for every test, so it's a bit of duplicated code. not a big deal, and not targeted specifically at this PR. just a thought.

**Review by azhou-determined — APPROVED** [review]

**Pull request merged**